File: server/chromium_dev.py
"""
Isolated Chromium for local server debugging — no daily-browser cache or profile.

Browsers live under server/.playwright-browsers (Playwright download).
Each launch uses a fresh ephemeral profile under server/.chromium-profiles/.
"""
from __future__ import annotations

import json
import logging
import os
import shutil
import subprocess
import sys
import threading
from pathlib import Path
from typing import Callable, Optional, Tuple

from . import config

logger = logging.getLogger('Rianell')

CHROMIUM_SCRIPT = Path(__file__).resolve().parent / 'scripts' / 'chromium-dev.mjs'
CHROMIUM_BROWSERS_DIR = Path(__file__).resolve().parent / '.playwright-browsers'
CHROMIUM_PROFILES_DIR = Path(__file__).resolve().parent / '.chromium-profiles'


def _node_executable() -> Optional[str]:
    return shutil.which('node')


def chromium_env() -> dict:
    env = os.environ.copy()
    env['PLAYWRIGHT_BROWSERS_PATH'] = str(CHROMIUM_BROWSERS_DIR)
    env['RIANELL_CHROMIUM_PROFILES'] = str(CHROMIUM_PROFILES_DIR)
    return env


def chromium_status() -> dict:
    """Return install paths and whether Chromium is present."""
    node = _node_executable()
    if not node or not CHROMIUM_SCRIPT.is_file():
        return {
            'installed': False,
            'nodeAvailable': bool(node),
            'scriptPresent': CHROMIUM_SCRIPT.is_file(),
            'browsersPath': str(CHROMIUM_BROWSERS_DIR),
            'profilesPath': str(CHROMIUM_PROFILES_DIR),
            'executable': None,
        }
    try:
        result = subprocess.run(
            [node, str(CHROMIUM_SCRIPT), 'status'],
            env=chromium_env(),
            cwd=str(config.PROJECT_ROOT),
            capture_output=True,
            text=True,
            timeout=30,
            check=False,
        )
        if result.returncode == 0 and result.stdout.strip():
            data = json.loads(result.stdout)
            data['nodeAvailable'] = True
            data['scriptPresent'] = True
            return data
    except (json.JSONDecodeError, subprocess.TimeoutExpired, OSError) as exc:
        logger.debug('chromium_status failed: %s', exc)
    return {
        'installed': _find_chromium_executable() is not None,
        'nodeAvailable': True,
        'scriptPresent': True,
        'browsersPath': str(CHROMIUM_BROWSERS_DIR),
        'profilesPath': str(CHROMIUM_PROFILES_DIR),
        'executable': _find_chromium_executable(),
    }


def _find_chromium_executable() -> Optional[str]:
    root = CHROMIUM_BROWSERS_DIR
    if not root.is_dir():
        return None
    if sys.platform == 'win32':
        for exe in root.rglob('chrome.exe'):
            parts = exe.parts
            if any(p.startswith('chrome-win') for p in parts):
                return str(exe)
        return None
    if sys.platform == 'darwin':
        for app in root.rglob('Chromium'):
            if app.is_file() and 'MacOS' in app.parts:
                return str(app)
        return None
    for exe in root.rglob('chrome'):
        if exe.is_file() and 'chrome-linux' in exe.parts:
            return str(exe)
    return None


def is_chromium_installed() -> bool:
    return bool(chromium_status().get('installed'))


def install_chromium(
    on_progress: Optional[Callable[[str], None]] = None,
) -> Tuple[bool, str]:
    """Download Chromium via Playwright into server/.playwright-browsers."""
    node = _node_executable()
    if not node:
        return False, 'Node.js was not found on PATH. Install Node 20+ from nodejs.org.'
    if not CHROMIUM_SCRIPT.is_file():
        return False, f'Missing launcher script: {CHROMIUM_SCRIPT}'

    CHROMIUM_BROWSERS_DIR.mkdir(parents=True, exist_ok=True)

    if on_progress:
        on_progress('Downloading Chromium (Playwright)… this may take a few minutes.')

    logger.info('Installing dev Chromium into %s', CHROMIUM_BROWSERS_DIR)
    try:
        proc = subprocess.run(
            [node, str(CHROMIUM_SCRIPT), 'install'],
            env=chromium_env(),
            cwd=str(config.PROJECT_ROOT),
            capture_output=True,
            text=True,
            timeout=900,
            check=False,
        )
    except subprocess.TimeoutExpired:
        return False, 'Chromium install timed out after 15 minutes.'
    except OSError as exc:
        return False, str(exc)

    if proc.returncode != 0:
        detail = (proc.stderr or proc.stdout or '').strip()[:500]
        return False, detail or f'Install failed (exit {proc.returncode}). Run npm ci at repo root first.'

    if is_chromium_installed():
        exe = _find_chromium_executable()
        msg = f'Chromium installed under server/.playwright-browsers'
        if exe:
            msg += f'\n{exe}'
        logger.info(msg.replace('\n', ' | '))
        return True, msg
    return False, 'Install finished but Chromium executable was not found.'


def launch_clean_chromium(url: str, watch_reload: bool = True) -> Tuple[bool, str]:
    """Open url in a detached, ephemeral-profile Chromium window."""
    node = _node_executable()
    if not node:
        return False, 'Node.js was not found on PATH.'
    if not CHROMIUM_SCRIPT.is_file():
        return False, f'Missing launcher script: {CHROMIUM_SCRIPT}'
    if not is_chromium_installed():
        return False, (
            'Chromium is not installed yet. Use "Download Chromium" in the server dashboard '
            'or run: python -m server.chromium_dev install'
        )

    CHROMIUM_PROFILES_DIR.mkdir(parents=True, exist_ok=True)
    cmd = [node, str(CHROMIUM_SCRIPT), 'launch', '--url', url]
    if not watch_reload:
        cmd.append('--no-watch-reload')
    logger.info('Launching clean Chromium: %s', url)

    try:
        kwargs = {
            'env': chromium_env(),
            'cwd': str(config.PROJECT_ROOT),
            'stdin': subprocess.DEVNULL,
        }
        if sys.platform == 'win32':
            subprocess.Popen(
                cmd,
                creationflags=subprocess.CREATE_NEW_PROCESS_GROUP | subprocess.DETACHED_PROCESS,  # type: ignore[attr-defined]
                close_fds=True,
                **kwargs,
            )
        else:
            subprocess.Popen(cmd, start_new_session=True, **kwargs)
    except OSError as exc:
        logger.error('Failed to launch Chromium: %s', exc)
        return False, str(exc)

    return True, (
        f'Clean Chromium opened at {url}'
        + (' (live reload via /api/reload enabled)' if watch_reload else '')
    )


def install_chromium_async(
    on_done: Callable[[bool, str], None],
    on_progress: Optional[Callable[[str], None]] = None,
) -> None:
    def worker():
        ok, msg = install_chromium(on_progress=on_progress)
        on_done(ok, msg)

    threading.Thread(target=worker, daemon=True).start()


def cli_main(argv: Optional[list] = None) -> int:
    import argparse

    parser = argparse.ArgumentParser(
        description='Dev Chromium for Rianell local server (isolated cache/profile).',
    )
    parser.add_argument(
        'command',
        nargs='?',
        choices=('status', 'install', 'launch'),
        default='status',
    )
    parser.add_argument('--url', default=f'http://127.0.0.1:{config.PORT}/')
    args = parser.parse_args(argv)

    if args.command == 'status':
        print(json.dumps(chromium_status(), indent=2))
        return 0 if is_chromium_installed() else 1
    if args.command == 'install':
        ok, msg = install_chromium(on_progress=print)
        print(msg)
        return 0 if ok else 1
    if args.command == 'launch':
        ok, msg = launch_clean_chromium(args.url)
        print(msg)
        return 0 if ok else 1
    return 2


if __name__ == '__main__':
    raise SystemExit(cli_main())

File: server/scripts/chromium-dev.mjs
#!/usr/bin/env node
/**
 * Dev Chromium for the Rianell local server — isolated from daily-use browsers.
 * Browsers: server/.playwright-browsers (PLAYWRIGHT_BROWSERS_PATH)
 * Ephemeral profiles: server/.chromium-profiles/run-*
 */
import { spawnSync } from 'node:child_process';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const SERVER_DIR = path.resolve(__dirname, '..');
const PROJECT_ROOT = path.resolve(SERVER_DIR, '..');

function browsersPath() {
  return process.env.PLAYWRIGHT_BROWSERS_PATH
    || path.join(SERVER_DIR, '.playwright-browsers');
}

function profilesPath() {
  return process.env.RIANELL_CHROMIUM_PROFILES
    || path.join(SERVER_DIR, '.chromium-profiles');
}

function playwrightEnv() {
  return {
    ...process.env,
    PLAYWRIGHT_BROWSERS_PATH: browsersPath(),
  };
}

async function loadPlaywright() {
  const candidates = [
    'playwright',
    path.join(PROJECT_ROOT, 'node_modules', 'playwright', 'index.mjs'),
    path.join(PROJECT_ROOT, 'benchmarks', 'node_modules', 'playwright', 'index.mjs'),
  ];
  for (const spec of candidates) {
    try {
      return await import(spec);
    } catch {
      // try next
    }
  }
  throw new Error(
    'Playwright is not installed. From repo root run: npm ci (or npm install in benchmarks/).',
  );
}

function findChromiumExecutable() {
  const root = browsersPath();
  if (!fs.existsSync(root)) return null;
  if (process.platform === 'win32') {
    const hits = [];
    walk(root, (p) => {
      if (p.endsWith('chrome.exe') && (p.includes('chrome-win') || p.includes('chrome-win64'))) hits.push(p);
    });
    return hits[0] || null;
  }
  if (process.platform === 'darwin') {
    let found = null;
    walk(root, (p) => {
      if (p.endsWith('Chromium.app/Contents/MacOS/Chromium')) found = p;
    });
    return found;
  }
  let found = null;
  walk(root, (p) => {
    if (p.endsWith(path.join('chrome-linux', 'chrome'))) found = p;
  });
  return found;
}

function walk(dir, fn) {
  for (const name of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, name.name);
    if (name.isDirectory()) walk(full, fn);
    else fn(full);
  }
}

function parseArgs(argv) {
  const out = {
    command: argv[0] || 'status',
    url: 'http://127.0.0.1:8080/',
    watchReload: true,
  };
  for (let i = 1; i < argv.length; i += 1) {
    const a = argv[i];
    if (a === '--url' && argv[i + 1]) {
      out.url = argv[++i];
    } else if (a === '--no-watch-reload') {
      out.watchReload = false;
    } else if (a === '--watch-reload') {
      out.watchReload = true;
    } else if (a.startsWith('http://') || a.startsWith('https://')) {
      out.url = a;
    }
  }
  return out;
}

/** Loopback hosts where the Python dev server exposes /api/reload (same gate as PWA index.html). */
function isDevReloadHost(hostname) {
  const h = String(hostname || '').toLowerCase();
  return h === 'localhost' || h === '127.0.0.1' || h === '[::1]';
}

function parseSseEvents(buffer) {
  const events = [];
  let rest = buffer;
  let sep;
  while ((sep = rest.indexOf('\n\n')) >= 0) {
    const block = rest.slice(0, sep);
    rest = rest.slice(sep + 2);
    for (const line of block.split('\n')) {
      if (!line.startsWith('data: ')) continue;
      try {
        events.push(JSON.parse(line.slice(6)));
      } catch {
        // ignore malformed chunks
      }
    }
  }
  return { events, rest };
}

/**
 * Subscribe to the server's SSE reload stream and invoke onReload for each push.
 * Mirrors connectToReloadStream() in apps/pwa-webapp/app.js (localhost dev only).
 */
async function watchDevReloadStream(pageUrl, onReload) {
  const origin = new URL(pageUrl).origin;
  const host = new URL(pageUrl).hostname;
  if (!isDevReloadHost(host)) {
    console.log(`Live reload watcher skipped (non-loopback host: ${host})`);
    return;
  }

  let attempt = 0;
  while (true) {
    try {
      const res = await fetch(`${origin}/api/reload`, {
        headers: {
          Accept: 'text/event-stream',
          'Cache-Control': 'no-cache',
        },
      });
      if (!res.ok || !res.body) {
        throw new Error(`SSE HTTP ${res.status}`);
      }

      attempt = 0;
      console.log(`Live reload: connected to ${origin}/api/reload`);

      const reader = res.body.getReader();
      const decoder = new TextDecoder();
      let buf = '';

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        buf += decoder.decode(value, { stream: true });
        const parsed = parseSseEvents(buf);
        buf = parsed.rest;
        for (const data of parsed.events) {
          if (data && data.type === 'reload') {
            console.log('Live reload: server push — refreshing page…');
            await onReload();
          }
        }
      }
    } catch (err) {
      attempt += 1;
      const delay = Math.min(2000 * 2 ** (attempt - 1), 16000);
      console.warn(`Live reload: reconnecting in ${delay}ms (${err?.message || err})`);
      await new Promise((resolve) => setTimeout(resolve, delay));
    }
  }
}

function cmdInstall() {
  fs.mkdirSync(browsersPath(), { recursive: true });
  console.log(`Installing Chromium into ${browsersPath()} …`);
  const npx = process.platform === 'win32' ? 'npx.cmd' : 'npx';
  const result = spawnSync(
    npx,
    ['playwright', 'install', 'chromium'],
    {
      cwd: PROJECT_ROOT,
      env: playwrightEnv(),
      stdio: 'inherit',
      shell: process.platform === 'win32',
    },
  );
  if (result.status !== 0) {
    process.exit(result.status || 1);
  }
  const exe = findChromiumExecutable();
  if (!exe) {
    console.error('Install finished but Chromium executable was not found.');
    process.exit(1);
  }
  console.log(`Chromium ready: ${exe}`);
}

async function cmdLaunch(url, options = {}) {
  const watchReload = options.watchReload !== false;
  fs.mkdirSync(profilesPath(), { recursive: true });
  const sessionDir = fs.mkdtempSync(path.join(profilesPath(), 'run-'));
  const pw = await loadPlaywright();
  const executablePath = findChromiumExecutable() || undefined;
  if (!executablePath) {
    console.error('Chromium is not installed. Run: node server/scripts/chromium-dev.mjs install');
    process.exit(1);
  }

  console.log(`Launching clean Chromium → ${url}`);
  console.log(`Profile (ephemeral): ${sessionDir}`);
  if (watchReload && isDevReloadHost(new URL(url).hostname)) {
    console.log('Live reload: enabled (same /api/reload stream as “Open app in browser”)');
  }

  const browser = await pw.chromium.launch({
    headless: false,
    executablePath,
    args: [
      '--disable-sync',
      '--no-first-run',
      '--disable-default-apps',
      '--disable-extensions',
      '--disable-features=TranslateUI',
      '--disable-component-update',
    ],
  });

  const context = await browser.newContext({
    ignoreHTTPSErrors: true,
    viewport: null,
  });
  context.setDefaultNavigationTimeout(120_000);
  await context.addInitScript(() => {
    window.__rianellExternalReloadWatcher = true;
  });
  const page = await context.newPage();
  await page.goto(url, { waitUntil: 'domcontentloaded' });

  if (watchReload) {
    watchDevReloadStream(url, async () => {
      try {
        await page.reload({ waitUntil: 'domcontentloaded' });
      } catch {
        // ignore reload races while the tab is closing
      }
    });
  }

  const cleanup = () => {
    try {
      fs.rmSync(sessionDir, { recursive: true, force: true });
    } catch {
      // ignore
    }
  };

  browser.on('disconnected', () => {
    cleanup();
    process.exit(0);
  });

  process.on('SIGINT', async () => {
    try {
      await browser.close();
    } catch {
      // ignore
    }
    cleanup();
    process.exit(0);
  });
}

function cmdStatus() {
  const exe = findChromiumExecutable();
  const payload = {
    installed: !!exe,
    browsersPath: browsersPath(),
    profilesPath: profilesPath(),
    executable: exe,
    platform: process.platform,
  };
  console.log(JSON.stringify(payload, null, 2));
  process.exit(exe ? 0 : 1);
}

async function main() {
  const { command, url, watchReload } = parseArgs(process.argv.slice(2));
  switch (command) {
    case 'install':
      cmdInstall();
      break;
    case 'launch':
      await cmdLaunch(url, { watchReload });
      break;
    case 'status':
      cmdStatus();
      break;
    default:
      console.error(`Unknown command: ${command}. Use install | launch | status`);
      process.exit(2);
  }
}

main().catch((err) => {
  console.error(err?.stack || err);
  process.exit(1);
});
